cpp/src/strings/search/contains_multiple.cu

```diff
@@ -0,0 +1,313 @@
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/logical.h>
+#include <thrust/sequence.h>
+#include <thrust/unique.h>
+
+#include <vector>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+namespace {
+
+/**
+ * @brief Threshold to decide on using string or warp parallel functions.
+ *
+ * If the average byte length of a string in a column exceeds this value then
+ * a warp-parallel function is used.
+ */
+constexpr size_type AVG_CHAR_BYTES_THRESHOLD = 64;
+
+/**
+ * @brief Kernel for finding multiple targets in each row of input strings
+ *
+ * The d_first_bytes is sorted and unique so the d_indices and d_offsets
+ * are used to map the corresponding character to its d_targets entry.
+ *
+ * Example
+ * d_targets = ["foo", "hello", "world", "hi"]
+ *  - sorted first-chars: ['f','h','h','w']
+ * d_indices = [0, 3, 1, 2]
+ * d_first_bytes = ['f', 'h', 'w']   (unique)
+ * d_offsets = [0, 1, 3]
+ * unique_count = 3
+ *
+ * If 'h' is found, lower_bound produces pos=1 in d_first_bytes.
+ * This corresponds to d_offset[1]==1 which has two values:
+ * - (d_offsets[2] - d_offsets[1]) = (3 - 1) = 2.
+ * Set map_idx = d_offsets[1] = 1 and the two targets to check are sequential
+ * in the d_indices array:
+ * - tgt1_idx = d_indices[map_idx]   = 3 --> d_targets[3] == 'hi'
+ * - tgt2_idx = d_indices[map_idx+1] = 1 --> d_targets[1] == 'hello'
+ * The logic now only needs to check for either of these 2 targets.
+ *
+ * This kernel works in either row-per-string or warp-per-string depending
+ * on the template parameter. If tile_size==1, then this kernel executes as
+ * a row-per-string. If tile_size=32, the it executes as a warp-per-string.
+ * No other options are supported for now.
+ *
+ * @tparam tile_size Number of threads per string
+ * @param d_strings Input strings
+ * @param d_targets Target strings to search within input strings
+ * @param d_first_bytes Sorted, unique list of first bytes of the target strings
+ * @param d_indices Indices to map sorted d_first_bytes to d_targets
+ * @param d_offsets Offsets to map d_indices to d_targets
+ * @param unique_count Number of unique values in d_first_bytes (and d_offsets)
+ * @param working_memory Global memory to use if shared-memory is too small
+ * @param d_results Bool results for each target within each string row
+ */
+template <cudf::thread_index_type tile_size>
+CUDF_KERNEL void multi_contains_kernel(column_device_view const d_strings,
+                                       column_device_view const d_targets,
+                                       u_char const* d_first_bytes,
+                                       size_type const* d_indices,
+                                       size_type const* d_offsets,
+                                       size_type unique_count,
+                                       bool* working_memory,
+                                       cudf::device_span<bool*> d_results)
```

Comment:
Not a hard requirement, using cooperative groups can simplify the logic here since there will be no more branching based on the CG size.

Reply:
Ok, I think this worth doing. This would make it even easier to try out different tile-sizes.